### PR TITLE
Expose the ResilienceHandler type.

### LINKED ISSUE #4759: [API Proposal]: Expose Resilience HTTP Message Handler
### Background and motivation

I'm attempting to migrate from `Microsoft.Extensions.Http.Polly` to `Microsoft.Extensions.Http.Resilience` (see https://github.com/dotnet/extensions/issues/4717).

However, in my code base there are some scenarios where I want to bypass `IHttpClientBuilder` and instantiate an HTTP delegating message handler directly. The former library exposes a public [`PolicyHttpMessageHandler`](https://github.com/dotnet/aspnetcore/blob/v8.0.0/src/HttpClientFactory/Polly/src/PolicyHttpMessageHandler.cs) for this purpose. Can we consider exposing [this library's handler](https://github.com/dotnet/extensions/blob/b31f7e964180e45b9b0855f24b91de024e0a438f/src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/ResilienceHandler.cs)?

To motivate, I'll illustrate two reasons I have found for bypassing the builder.

1) To prevent duplicate handlers from being registered when `AddResilienceHandler` is called more than once. 
2) To configure multiple resilience handlers at once using the client name.

These can both be accomplished with a custom implementation of `IConfigureNamedOptions<HttpClientFactoryOptions>`:

```csharp
class ConfigureHttpResilience : IConfigureNamedOptions<HttpClientFactoryOptions>
{
        public void Configure(string? name, HttpClientFactoryOptions options)
        {
            if (name != null)
            {
                // Create a custom resilience pipeline using the client name.
                var pipeline = GetPipeline(name);

                // Can't do this since ResilienceHandler is not public!
                options.HttpMessageHandlerBuilderActions.Add(
                    builder => builder.AdditionalHandlers.Add(new ResilienceHandler(x => pipeline)));
            }
        }
}
```

```csharp

public static IServiceCollection AddHttpResilience(this IServiceCollection services)
{
    // "Try" prevents duplicates if this method is called more than once.
    services.TryAddEnumerable(
        ServiceDescriptor.Singleton<IConfigureOptions<HttpClientFactoryOptions>, ConfigureHttpResilience>());

    return services;
}
```

### API Proposal

```csharp
public class ResilienceHttpMessageHandler : DelegatingHandler
{
    public ResilienceHttpMessageHandler(ResiliencePipeline<HttpResponseMessage> pipeline)
    {
        throw new NotImplementedException();
    }

    public ResilienceHttpMessageHandler(Func<HttpRequestMessage, ResiliencePipeline<HttpResponseMessage>> pipelineProvider)
    {
        throw new NotImplementedException();
    }

    protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        throw new NotImplementedException();
    }
}
```

### API Usage

```csharp
options.HttpMessageHandlerBuilderActions.Add(x => x.AdditionalHandlers.Add(new ResilienceHttpMessageHandler(pipeline)));
```

### Alternative Designs

_No response_

### Risks

Future changes to the handler's public constructors are more likely to be breaking.

### PR DESCRIPTION
**Pull request by geeknoid**
Fixes #4759

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4858)

### CODE REVIEW

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs [hunk 1]
```diff
@@ -2,29 +2,47 @@
...
+using Microsoft.Extensions.Http.Resilience.Internal;
+using Microsoft.Shared.DiagnosticIds;
+using Microsoft.Shared.Diagnostics;
 using Polly;
 
-namespace Microsoft.Extensions.Http.Resilience.Internal;
+namespace Microsoft.Extensions.Http.Resilience;
 
 /// <summary>
 /// Base class for resilience handler, i.e. handlers that use resilience strategies  to send the requests.
```
Comment (martintmk) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs:18]:
```suggestion
/// Base class for resilience handler, i.e. handlers that use resilience strategies to send the requests.
```

nit

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs [hunk 2]
```diff
@@ -2,29 +2,47 @@
...
+public class ResilienceHandler : DelegatingHandler
 {
     private readonly Func<HttpRequestMessage, ResiliencePipeline<HttpResponseMessage>> _pipelineProvider;
 
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResilienceHandler"/> class.
+    /// </summary>
+    /// <param name="pipelineProvider">The pipeline provider that supplies pipelines in response to an http message.</param>
+    /// <exception cref="ArgumentNullException">If <paramref name="pipelineProvider"/> is <see langword="null"/>.</exception>
     public ResilienceHandler(Func<HttpRequestMessage, ResiliencePipeline<HttpResponseMessage>> pipelineProvider)
```
Comment (martintmk) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs:30]: How about alternative constructor for simple scenarios that just accepts `ResiliencePipeline`?

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs [hunk 3]
```diff
@@ -2,29 +2,47 @@
...
 
-    /// <inheritdoc/>
+    /// <summary>
+    /// Sends an HTTP request to the inner handler to send to the server as an asynchronous operation.
+    /// </summary>
+    /// <param name="request">The HTTP request message to send to the server.</param>
+    /// <param name="cancellationToken">A cancellation token to cancel operation.</param>
+    /// <returns>The task object representing the asynchronous operation.</returns>
+    /// <exception cref="ArgumentNullException">If <paramref name="request"/> is <see langword="null"/>.</exception>
     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
```
Comment (martintmk) [inline src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs:42]: Would `inheritdoc` work here?

Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs:42]: Nope, inheritdoc only works for one "region" of documentation, so it wouldn't discover the docs from the other repos.

File: src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs (hunk 2 above)
Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/ResilienceHandler.cs:30]: Done.

**Review by martintmk — APPROVED** [review]

**Pull request merged**